src/test/java/com/amazon/ion/util/ZstdStreamInterceptorTest.java

```diff
@@ -0,0 +1,379 @@
+
+        private final int length;
+
+        LengthTooLongInterceptor(int length) {
+            this.length = length;
+        }
+
+        @Override
+        public String formatName() {
+            return null;
+        }
+
+        @Override
+        public int headerMatchLength() {
+            return length;
+        }
+
+        @Override
+        public boolean matchesHeader(byte[] candidate, int offset, int length) {
+            return Assertions.fail("This method should be unreachable.");
+        }
+
+        @Override
+        public InputStream newInputStream(InputStream interceptedStream) {
+            return Assertions.fail("This method should be unreachable.");
+        }
+    }
+
+    @ParameterizedTest
+    @EnumSource(ZstdStream.class)
+    public void notInterceptedWhenStreamLengthIsLessThanHeaderLength(ZstdStream stream) throws IOException {
+        IonReaderBuilder builder = IonReaderBuilder.standard()
+            // The LengthTooLongInterceptor should be skipped, then the ZstdStreamInterceptor matched.
+            .addInputStreamInterceptor(new LengthTooLongInterceptor(1000)) // None of the test data is this long.
+            .addInputStreamInterceptor(new ZstdStreamInterceptor());
+        try (IonReader reader = stream.newReader(builder)) {
+            assertEquals(IonType.INT, reader.next());
+            assertEquals(123, reader.intValue());
+        }
+    }
+
+    @ParameterizedTest
+    @EnumSource(ZstdStream.class)
+    public void expectFailureWhenHeaderLengthIsInvalid(ZstdStream stream) {
+        IonReaderBuilder builder = IonReaderBuilder.standard()
+            // This header length is invalid because an array of that size cannot be allocated.
+            .addInputStreamInterceptor(new LengthTooLongInterceptor(Integer.MAX_VALUE))
+            .addInputStreamInterceptor(new ZstdStreamInterceptor());
+        assertThrows(IonException.class, () -> stream.newReader(builder));
+    }
+
+    private static class OneBytePerReadInputStream extends InputStream {
+
+        private final InputStream delegate;
+
+        OneBytePerReadInputStream(InputStream delegate) {
+            this.delegate = delegate;
+        }
+
+        @Override
+        public int read() throws IOException {
+            return delegate.read();
+        }
+
+        @Override
+        public int read(byte[] bytes, int off, int len) throws IOException {
+            int b = delegate.read();
+            if (b < 0) {
+                return -1;
+            }
+            bytes[off] = (byte) b;
+            return 1;
+        }
+    }
+
+    @Test
+    public void headerRequiresMultipleInputStreamReads() throws IOException {
+        IonReaderBuilder builder = IonReaderBuilder.standard()
+            .addInputStreamInterceptor(new ZstdStreamInterceptor());
+        try (IonReader reader = builder.build(new OneBytePerReadInputStream(new ByteArrayInputStream(ZstdStream.BINARY_BYTES)))) {
```

Comment:
```suggestion
        try (IonReader reader = builder.build(new OneBytePerReadInputStream(stream(binaryBytes())))) {
```

Assuming you take the ZstdStream suggestion.